anonlinkclient/cli.py

```diff
@@ -331,36 +336,75 @@ def create(name, project, apikey, output, threshold, server, retry_multiplier, r
         json.dump(response, output)
 
 
+#todo
+#@click.option('--upload-to', help="User supplied object store path. e.g. 's3://bucket/path'", default=None)
+
 @cli.command('upload', short_help='upload hashes to entity service')
-@click.argument('clk_json', type=click.File('rb'))
+@click.argument('clk_json', type=click.Path(exists=True, dir_okay=False))
 @click.option('--project', help='Project identifier')
 @click.option('--apikey', help='Authentication API key for the server.')
 @click.option('-o', '--output', type=click.File('w'), default='-')
 @click.option('--blocks', help='Generated blocks JSON file', type=click.File('rb'))
 @add_options(rest_client_option)
+@click.option('--profile', help="AWS profile to use if uploading to own S3 Bucket", default=None)
 @verbose_option
-def upload(clk_json, project, apikey, output, blocks, server, retry_multiplier, retry_max_exp, retry_stop, verbose):
-    """Upload CLK data to entity matching server.
+def upload(clk_json, project, apikey, output, blocks, server, retry_multiplier, retry_max_exp, retry_stop, profile, verbose):
+    """Upload CLK data to the Anonlink Entity server.
 
     Given a json file containing hashed clk data as CLK_JSON, upload to
     the entity resolution service.
 
-    Use "-" to read from stdin.
+    The following environment variables can be used to override default behaviour:
+
+    * UPLOAD_OBJECT_STORE_SERVER
+
     """
     msg = 'CLK and Blocks' if blocks else 'CLK'
+
     if verbose:
-        log("Uploading CLK data from {}".format(clk_json.name))
+        log("Uploading CLK data from {}".format(clk_json))
         log("Project ID: {}".format(project))
         log("Uploading {} data to the server".format(msg))
 
     rest_client = create_rest_client(server, retry_multiplier, retry_max_exp, retry_stop, verbose)
 
+    if verbose: log("Fetching temporary credentials")
+    res = rest_client.get_temporary_objectstore_credentials(project, apikey)
+    credentials = res['credentials']
+    upload_info = res['upload']
+
+    endpoint = os.getenv('UPLOAD_OBJECT_STORE_SERVER', upload_info['endpoint'])
+
+    object_store_credential_providers = []
+    if profile is not None:
+        object_store_credential_providers.append(FileAWSCredentials(profile=profile))
+
+    object_store_credential_providers.append(
+        Static(access_key=credentials['AccessKeyId'],
+               secret_key=credentials['SecretAccessKey'],
+               token=credentials['SessionToken']))
+
+    mc = Minio(
+        endpoint,
+        credentials=Credentials(provider=Chain(object_store_credential_providers)),
+        region='us-east-1',
+        secure=upload_info['secure']
+    )
+
+    if verbose:
+        log('Checking we have permission to upload')
+    mc.put_object(upload_info['bucket'], upload_info['path'] + "/upload-test", io.BytesIO(b"something"), length=9)
```

Comment:
why the length is 9 here?

Reply:
Is this some dummy code to upload "something" to minio?

Reply:
Yup :-) and the length of "something" is 9.

It *might* be a good idea to keep though - it will fail if there is any issue with the credentials or server before we start processing the encodings or blocks to get them ready for upload.

Reply:
Got it. Nice :)